### PR TITLE
Fix the warnings with Constant and subfunctions.

### PR DESCRIPTION
**Pull request by Ig-dolci**
This PR updates the firedrake tests with the correct definitions.

- Move from `Connstant(any_value, domain=mesh)` to `Functtion(R).assign(any_value)`, where `R = FunctionSpace(mesh, "R", 0)`. 
- Move from `f.split` to `f.sub`.

### CODE REVIEW

File: tests/firedrake_adjoint/test_disk_checkpointing.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def adjoint_example(fine, coarse):
     # AssembleBlock
     m = assemble(interpolate(sin(4*pi*x)*cos(4*pi*y), cg_space))
 
-    u, v = w.split()
+    u, v = w.sub(0), w.sub(1)
```
Comment (dham) [inline tests/firedrake_adjoint/test_disk_checkpointing.py:22]:
```suggestion
    u, v = w.subfunctions
```

File: tests/firedrake_adjoint/test_projection.py [hunk 2]
```diff
@@ -135,36 +135,42 @@ def test_project_nonlin_changing():
...
     u = Function(V)
-    c = Constant(1., domain=mesh)
+    c = Function(R).assign(1.)
     u.project(u+c)
     J = assemble(u**2*dx)
     rf = ReducedFunctional(J, Control(c))
     h = Constant(0.1)
-    assert taylor_test(rf, Constant(2., domain=mesh), h)
+    m = Function(R).assign(2.)
+    assert taylor_test(rf, m, h)
```
Comment (dham) [inline tests/firedrake_adjoint/test_projection.py:146]: Why the additional variable?

File: tests/firedrake_adjoint/test_projection.py [hunk 3]
```diff
@@ -135,36 +135,42 @@ def test_project_nonlin_changing():
...
-    c = Constant(1., domain=mesh)
+    c = Function(R).assign(1.)
     project(u+c, u)
     project(u+c*u**2, u)
     J = assemble(u**2*dx)
     rf = ReducedFunctional(J, Control(c))
     h = Constant(0.1)
-    assert taylor_test(rf, Constant(3., domain=mesh), h)
+    m = Function(R).assign(3.)
+    assert taylor_test(rf, m, h)
```
Comment (dham) [inline tests/firedrake_adjoint/test_projection.py:160]: no additional variable.

File: tests/firedrake_adjoint/test_reduced_functional.py [hunk 4]
```diff
@@ -9,21 +9,24 @@
...
+    bc = DirichletBC(V, bv, "on_boundary")
 
     F = inner(grad(u), grad(v))*dx - f**2*v*dx
     solve(F == 0, u, bc)
 
     J = assemble(c**2*u*dx)
     Jhat = ReducedFunctional(J, Control(c))
-    assert taylor_test(Jhat, c, Constant(1, domain=mesh)) > 1.9
+    m = Function(R).assign(1)
+    assert taylor_test(Jhat, c, m) > 1.9
```
Comment (dham) [inline tests/firedrake_adjoint/test_reduced_functional.py:29]: no additional variable.

File: tests/firedrake_adjoint/test_solving.py [hunk 5]
```diff
@@ -9,14 +9,15 @@
...
+    R = FunctionSpace(mesh, "R", 0)
     f = Function(V)
     f.vector()[:] = 1
 
     u = TrialFunction(V)
     u_ = Function(V)
     v = TestFunction(V)
-    bc = DirichletBC(V, Constant(1, domain=mesh), "on_boundary")
+    bv = Function(R).assign(1)
+    bc = DirichletBC(V, bv, "on_boundary")
```
Comment (dham) [inline tests/firedrake_adjoint/test_solving.py:20]: no additional variable.

File: tests/firedrake_adjoint/test_solving.py [hunk 6]
```diff
@@ -60,13 +61,14 @@ def test_nonlinear_problem():
...
-
+    R = FunctionSpace(mesh, "R", 0)
     f = Function(V)
     f.vector()[:] = 1
 
     u = Function(V)
     v = TestFunction(V)
-    bc = DirichletBC(V, Constant(1, domain=mesh), "on_boundary")
+    bv  = Function(R).assign(1)
+    bc = DirichletBC(V, bv, "on_boundary")
```
Comment (dham) [inline tests/firedrake_adjoint/test_solving.py:71]: and here.

File: tests/firedrake_adjoint/test_split.py [hunk 7]
```diff
@@ -108,23 +109,26 @@ def test_fn_split_no_annotate(Z):
...
 def test_merge_blocks():
     mesh = UnitSquareMesh(1,1)
     V = FunctionSpace(mesh, 'CG', 1)
+    R = FunctionSpace(mesh, 'R', 0)
     W = V * V
     w = Function(W)
-    w1, w2 = w.split()
-    w1_const = Constant(0.1, domain=mesh)
-    w2_const = Constant(0.2, domain=mesh)
+    w1, w2 = w.sub(0), w.sub(1)
```
Comment (dham) [inline tests/firedrake_adjoint/test_split.py:122]:
```suggestion
    w1, w2 = w.subfunctions
```

File: tests/firedrake_adjoint/test_split.py [hunk 8]
```diff
@@ -108,23 +109,26 @@ def test_fn_split_no_annotate(Z):
...
+    w2_const = Function(R).assign(0.2)
     w1.project(w1_const)
     w2.project(w2_const)
     J = assemble(w1*w1*dx)
     c = Control(w1_const)
     rf = ReducedFunctional(J, c)
-    assert taylor_test(rf, Constant(0.3, domain=mesh), Constant(0.01, domain=mesh)) > 1.95
+    m = Function(R).assign(0.3)
+    h = Function(R).assign(0.01)
+    assert taylor_test(rf, m, h) > 1.95
```
Comment (dham) [inline tests/firedrake_adjoint/test_split.py:132]: don't introduce new variables.

**Review by dham — CHANGES_REQUESTED** [review]

File: tests/firedrake_adjoint/test_reduced_functional.py [hunk 9]
```diff
@@ -166,10 +168,11 @@ def test_mixed_boundary():
 def test_assemble_recompute():
     mesh = UnitSquareMesh(10, 10)
     V = FunctionSpace(mesh, "CG", 1)
+    R = FunctionSpace(mesh, "R", 0)
 
     f = Function(V)
     f.vector()[:] = 2
-    expr = Constant(assemble(f**2*dx), domain=mesh)
+    expr = Function(R).assign(assemble(f**2*dx))
```
Comment (Ig-dolci) [inline tests/firedrake_adjoint/test_reduced_functional.py:175]: This test fails if I set `Function(R, val=assemble(f**2*dx))`. So, I kept `Function(R).assign(assemble(f**2*dx))`.

File: tests/firedrake_adjoint/test_projection.py (hunk 2 above)
Comment (Ig-dolci) [reply tests/firedrake_adjoint/test_projection.py]: done.

File: tests/firedrake_adjoint/test_projection.py (hunk 3 above)
Comment (Ig-dolci) [reply tests/firedrake_adjoint/test_projection.py]: done.

File: tests/firedrake_adjoint/test_reduced_functional.py (hunk 4 above)
Comment (Ig-dolci) [reply tests/firedrake_adjoint/test_reduced_functional.py]: done.

File: tests/firedrake_adjoint/test_solving.py (hunk 5 above)
Comment (Ig-dolci) [reply tests/firedrake_adjoint/test_solving.py]: done.

File: tests/firedrake_adjoint/test_solving.py (hunk 6 above)
Comment (Ig-dolci) [reply tests/firedrake_adjoint/test_solving.py]: done.

**Review by dham — APPROVED** [review]

**Pull request merged**